nixos/modules/services/home-automation/zwave-js-server.nix

```diff
@@ -0,0 +1,140 @@
+with lib;
+
+let
+  cfg = config.services.zwave-js-server;
+  cacheDir = "/var/cache/zwave-js-server";
+  runtimeDir = "/run/zwave-js-server";
+  defaultConfig = {
+    storage = {
+      inherit cacheDir;
+    };
+  };
+  mergedConfigFile = "${runtimeDir}/config.json";
+in {
+
+  options.services.zwave-js-server = {
+    enable = mkEnableOption (mdDoc "Start the zwave-js-server on boot");
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.nodePackages."@zwave-js/server";
+      defaultText = literalExpression "pkgs.nodePackages.\"@zwave-js/server\"";
+      description = mdDoc ''
+        Package for zwave-js server
+      '';
+    };
+
+    port = mkOption {
+      type = types.int;
+      default = 3000;
+      description = mdDoc ''
+        Port for the server to listen on.
+      '';
+    };
+
+    serialPort = mkOption {
+      type = types.path;
+      description = mdDoc ''
+        Serial port device path for Z-Wave controller.
+      '';
+      example = "/dev/ttyUSB0";
+    };
+
+    user = mkOption {
+      type = types.str;
+      default = "zwavejs";
+      description = mdDoc ''
+        User name for the zwave-js-server user.
+      '';
+    };
+
+    group = mkOption {
+      type = types.str;
+      default = "zwavejs";
+      description = mdDoc ''
+        Group name for the zwave-js-server user.
+      '';
+    };
+
+    secretsConfigFile = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = mdDoc ''
+        JSON file containing secret keys. This file will be merged with the
+        module-generated config file (taking precedence).
+
+        Z-Wave keys can be generated with:
+
+          {command}`< /dev/urandom tr -dc A-F0-9 | head -c32 ;echo`
+
+        ::: {.warning}
+        A file in the nix store should not be used since it will be readable to
+        all users.
+        :::
+      '';
+      example = ''
+        {
+          "securityKeys": {
+            "S2_Unauthenticated": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
+            "S2_Authenticated": "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
+            "S2_AccessControl": "CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC",
+            "S0_Legacy": "DDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDD"
+          }
+        }
+      '';
+    };
+
+    extraConfig = mkOption {
+      type = types.attrs;
+      default = {};
+      description = mdDoc ''
+        Extra configuration to be merged into the generated config file.
+      '';
+      example = literalExpression ''
+        {
+          storage = {
+            throttle = "fast";
+          };
+        }
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.tmpfiles.rules = [
+      "d ${runtimeDir} 0700 ${cfg.user} ${cfg.group}"
+      "d ${cacheDir} 0700 ${cfg.user} ${cfg.group}"
+    ];
+
+    users.users."${cfg.user}" = {
+      isSystemUser = true;
+      inherit (cfg) group;
+      extraGroups = [ "dialout" ];
+    };
+
+    users.groups."${cfg.group}" = {};
```

Comment:
```
systemd.services.zwave-js-server.serviceConfig = {
  DynamicUser = true;
  User = "zwave-js";
  SupplementaryGroups = [ "dialout" ];
}

Reply:
I don't have much experience with writing modules, I'll take a look, thanks.

Reply:
The zigbee2mqtt module might be a good inspiration.

Reply:
Took a look at https://0pointer.net/blog/dynamic-users-with-systemd.html - seems like using DynamicUser should be possible, will give it a try!

Reply:
Seems to work, haven't done further hardening (e.g. device access) yet.